### PR TITLE
Fix for "Adding a new primary language on iOS 9 doesn't set that language as the search language"

### PR DESCRIPTION
**Pull request by montehurd**
https://phabricator.wikimedia.org/T146403

The reported issue was really hard to repro consistently, but I noticed some of the search lang bar layout/configuration bits were being called multiple times (iirc 3 times on first display of the search interface). 

I ended up pulling *all* of the search lang bar button logic out of the `WMFSearchViewController` and into its own controller - `WMFSearchLanguagesBarViewController`. I also switched it to use normal view containment to get the new controller's view into `WMFSearchViewController`'s view.

* haven't seen the iOS 9 issue again
* no more repeated calls
* methods in the new controller are explicit about what they do
* explicit delegate protocol for being notified when search lang changes

### CODE REVIEW

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    private var _currentlySelectedSearchLanguage: MWKLanguageLink?
+    private(set) var currentlySelectedSearchLanguage: MWKLanguageLink? {
+        get {
+            return _currentlySelectedSearchLanguage
+        }
+        set {
+            _currentlySelectedSearchLanguage = newValue
+            NSUserDefaults.wmf_userDefaults().wmf_setCurrentSearchLanguageDomain(newValue?.siteURL())
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: We should handle the optional here. I would recommend setting the current search language tot he primary language if this value is nil

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            delegate?.searchLanguagesBarViewController(self, didChangeCurrentlySelectedSearchLanguage: newValue!)
+            updateLanguageBarLanguageButtons()
+        }
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        _ = languageButtons.map{$0.tintColor = UIColor.wmf_blueTintColor()}
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: shouldn't this just be enumerate?

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                heightConstraint!.constant = 0
+                view.hidden = true
+            }else{
+                heightConstraint!.constant = 44
+                view.hidden = false
+            }
+        }
+    }
+
+    private var _currentlySelectedSearchLanguage: MWKLanguageLink?
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: confused at the difference between `_currentlySelectedSearchLanguage` and `selectedLanguage` - this logic should probably be rolled into `selectedLanguage`

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    private(set) var currentlySelectedSearchLanguage: MWKLanguageLink? {
+        get {
+            return _currentlySelectedSearchLanguage
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Introducing extra state - `currentlySelectedSearchLanguage` should never be different than `NSUserDefaults.wmf_userDefaults().wmf_currentSearchLanguageDomain`. This should just return/set that value.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    override func viewDidAppear(animated: Bool) {
+        super.viewDidAppear(animated)
+        selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged()
+        previousFirstLanguage = languageBarLanguages().first
+    }
+    
+    private func languageBarLanguages() -> [MWKLanguageLink] {
+        return MWKLanguageLinkController.sharedInstance().preferredLanguages.enumerate().flatMap{ $0.index < 3 ? $0.element : nil }
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: `preferredLanguages.prefix(3)` will grab the first 3 elements without flatmapping and is bounds safe

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    
+    override func viewWillAppear(animated: Bool) {
+        super.viewWillAppear(animated)
+        updateLanguageBarLanguageButtons()
+        hidden = !NSUserDefaults.wmf_userDefaults().wmf_showSearchLanguageBar()
+    }
+
+    override func viewDidAppear(animated: Bool) {
+        super.viewDidAppear(animated)
+        selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged()
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: No need to have a special method to do checking,  just call your logic to set it to the right language:
`updateLanguageBarLanguageButtons()`

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        previousFirstLanguage = languageBarLanguages().first
+    }
+    
+    private func languageBarLanguages() -> [MWKLanguageLink] {
+        return MWKLanguageLinkController.sharedInstance().preferredLanguages.enumerate().flatMap{ $0.index < 3 ? $0.element : nil }
+    }
+
+    private func selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged(){
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: As noted above, you should be able to remove this method all together

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        if(isEveryButtonUnselected() || isFirstLanguageDifferentFromLastTime()){
+            setLanguage(withSender:languageButtons.first!)
+        }
+    }
+    
+    private func isEveryButtonUnselected() -> Bool{
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Should be able to remove this one too…

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        for button in languageButtons {
+            if button.selected {
+                return false
+            }
+        }
+        return true
+    }
+
+    private func isFirstLanguageDifferentFromLastTime() -> Bool{
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: …And this one

File: Wikipedia/Code/SearchLanguagesBarViewController.swift [hunk 10]
```diff
@@ -0,0 +1,145 @@
+import Foundation
+
+@objc protocol SearchLanguagesBarViewControllerDelegate: class {
+    func searchLanguagesBarViewController(controller: SearchLanguagesBarViewController, didChangeCurrentlySelectedSearchLanguage language: MWKLanguageLink)
+}
+
+class SearchLanguagesBarViewController: UIViewController, WMFPreferredLanguagesViewControllerDelegate {
```
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: We should discus removing `WMF` from classes in defined in Swift - so far we have been prefixing our Swift classes.

I could go either way, but would be good to set a best practice here.

@joewalsh any thoughts?

**Comment by coreyfloyd** [comment]
@montehurd most of the refactoring looks good. Left a few comments.

In regards to the bug itself. I'm not sure we can close this issue unless we can do one of the following:
1. Find a logic error in the previous code where there was an opportunity to be in a bad state and show how the new logic handles it correctly.
Or
2. Determine some repro steps to cause the bug and show it now works with the new code.

I saw the notes about the repeated calls updating the search bar. While its nice to not call view update logic unless an update is actually needed, it's not uncommon for views to be updated multiple times before they are are displayed on screen and doesn't necessarily point to a bug.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 5 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Oh cool! Forgot it was bounds safe...

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 2 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Sure, if you feel strongly about it.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 6 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Part of what was confusing about the old logic was it was hard to follow why certain things were being set. Made this explicit so it was super clear why it was happening.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 10 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Cool! I had thought we had said we didn't have to prefix the Swift classes, but I kinda prefer prefixing them personally.

Comment (joewalsh) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: @coreyfloyd seems like what we should be doing is dropping the prefix and using @objc(WMFClassNameHere).

`@objc (WMFSearchLanguagesBarViewController) class SearchLanguagesBarViewController {`

Anything exposed to or used by Objective-C should have a prefix.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 2 above)
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: In this case, yes… one of the things about swift is that it is functional… when you want to transform an object not just iterate, it provides you a map function to make your intentions clear.

This case is just the opposite, you really just want to enumerate, and it would be best to use the method that clearly identifies that intent of just iterating.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 6 above)
Comment (coreyfloyd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: I can see the value in removing complexity , but looking at the fallout in my other comments, it seems the complexity has just moved. There are now 3 methods and an ivar where previously there was just one method. 

I'm not sure the trade off between calling a method multiple times is worth the added complexity of 3 methods and an ivar that we now have to keep in sync with user defaults.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 10 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Re-added the prefix.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 5 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Fixed.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 2 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Fair enough. I fell victim to "new shiney" syndrome...
http://www.mokacoding.com/blog/when-to-use-map-flatmap-for/

Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Fixed.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 1 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Fixed.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 3 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Fixed.

File: Wikipedia/Code/SearchLanguagesBarViewController.swift (hunk 4 above)
Comment (montehurd) [inline Wikipedia/Code/SearchLanguagesBarViewController.swift]: Fixed.

File: Wikipedia/Code/WMFSearchLanguagesBarViewController.swift [hunk 11]
```diff
@@ -24,13 +24,21 @@ class WMFSearchLanguagesBarViewController: UIViewController, WMFPreferredLanguag
         }
     }
 
-    private var _currentlySelectedSearchLanguage: MWKLanguageLink?
     private(set) var currentlySelectedSearchLanguage: MWKLanguageLink? {
```
Comment (coreyfloyd) [inline Wikipedia/Code/WMFSearchLanguagesBarViewController.swift]: Still have variable here… see line 33. 
To remove the backing store you should make this into a computed property. See:
http://www.codingexplorer.com/computed-properties-in-swift/

File: Wikipedia/Code/WMFSearchLanguagesBarViewController.swift [hunk 12]
```diff
@@ -71,7 +76,7 @@ class WMFSearchLanguagesBarViewController: UIViewController, WMFPreferredLanguag
 
     private func selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged(){
```
Comment (coreyfloyd) [inline Wikipedia/Code/WMFSearchLanguagesBarViewController.swift]: You shouldn't need this method anymore… just call setCurrentlySelectedLanguageToButtonLanguage() directly from viewDidAppear

File: Wikipedia/Code/WMFSearchLanguagesBarViewController.swift [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+class WMFSearchLanguagesBarViewController: UIViewController, WMFPreferredLanguagesViewControllerDelegate {
+    weak var delegate: WMFSearchLanguagesBarViewControllerDelegate?
+    
+    @IBOutlet private var languageButtons: [UIButton] = []
+    @IBOutlet private var otherLanguagesButton: UIButton?
+    @IBOutlet private var heightConstraint: NSLayoutConstraint?
+    
+    private var previousFirstLanguage: MWKLanguageLink?
```
Comment (coreyfloyd) [inline Wikipedia/Code/WMFSearchLanguagesBarViewController.swift]: removing that method also removes the need for this variable

File: Wikipedia/Code/WMFSearchLanguagesBarViewController.swift [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    private func isEveryButtonUnselected() -> Bool{
+        for button in languageButtons {
+            if button.selected {
+                return false
+            }
+        }
+        return true
+    }
+
+    private func isFirstLanguageDifferentFromLastTime() -> Bool{
```
Comment (coreyfloyd) [inline Wikipedia/Code/WMFSearchLanguagesBarViewController.swift]: It also removes the need for this function

**Comment by coreyfloyd** [comment]
@montehurd just left a few more comments… 

I saw you made some changes, but the backing ivar is still present and the extra logic for checking state is still here as well.  Not sure if you are finished yet or now. 

Also tests are failing - you should check out travis.

**Comment by montehurd** [comment]
@coreyfloyd oh sorry I have one more commit I haven't pushed yet - wanted to talk w josh first because if he's cool with only changing the selected lang when the previously selected lang is no longer in the top 3 I can simplify it even further...

**Comment by montehurd** [comment]
(that's why I threw the WIP tag on it last night)

**Comment by montehurd** [comment]
@coreyfloyd ok seems pretty solid now

**Comment by montehurd** [comment]
@coreyfloyd ok seems pretty solid now... tested quite a bit... fingers crossed...

**Comment by coreyfloyd** [comment]
👍
